rust/core-lib/src/plugins.rs

```diff
@@ -0,0 +1,401 @@
+        let key = (buffer_id.to_owned(), plugin_name.to_owned());
+        if self.running.contains_key(&key) {
+            print_err!("plugin {} already running for buffer {}", plugin_name, buffer_id);
+        }
+        let plugin = self.catalog.iter().find(|desc| desc.name == plugin_name)
+            .expect(&format!("no plugin found with name {}", plugin_name));
+
+        let me = self_ref.clone();
+        plugin.launch(self_ref, buffer_id, move |result| {
+            match result {
+                Ok(plugin_ref) => {
+                    plugin_ref.init_buf(buf_size, rev);
+                    me.lock().unwrap().running.insert(key, plugin_ref);
+                },
+                Err(_) => panic!("error handling is not implemented"),
+            }
+        });
+    }
+
+    /// Handle a request from a plugin.
+    fn handle_plugin_cmd(&self, cmd: PluginCommand, view_id: &ViewIdentifier) -> Option<Value> {
+        use self::PluginCommand::*;
+        match cmd {
+            LineCount => {
+                let n_lines = self.buffers.lock().editor_for_view(view_id).unwrap()
+                    .plugin_n_lines() as u64;
+                Some(serde_json::to_value(n_lines).unwrap())
+            },
+            SetFgSpans { start, len, spans, rev } => {
+                self.buffers.lock().editor_for_view_mut(view_id).unwrap()
+                    .plugin_set_fg_spans(start, len, spans, rev);
+                None
+            }
+            GetData { offset, max_size, rev } => {
+                self.buffers.lock().editor_for_view(view_id).unwrap()
+                .plugin_get_data(offset, max_size, rev)
+                .map(|data| Value::String(data))
+            }
+
+            Alert { msg } => {
+                self.buffers.lock().editor_for_view(view_id).unwrap()
+                .plugin_alert(&msg);
+                None
+            }
+        }
+    }
+}
+
+
+/// Starts a thread which collects editor updates and propagates them to plugins.
+///
+/// In addition to updates caused by user edits, updates can be caused by
+/// plugin edits. These updates arrive asynchronously. After being applied to
+/// the relevant buffer via an `Editor` instance, they need to be propagated
+/// back out to all interested plugins.
+///
+/// In order to avoid additional complexity in the model graph (e.g. giving each
+/// `Editor` a weak reference to the `PluginManager`) we instead give each
+/// `Editor` a tx end of an `mpsc::channel`. As plugin updates are generated,
+/// they are sent over this channel to a receiver running in another thread,
+/// which forwards them to interested plugins.
+pub fn start_update_thread<W: Write + Send + 'static>(
+    rx: mpsc::Receiver<(ViewIdentifier, PluginUpdate, usize)>,
+    plugins: Arc<Mutex<PluginManager<W>>>)
+{
+    thread::spawn(move ||{
+        loop {
+            let (view_id, update, undo_group) = rx.recv().unwrap();
```

Comment:
This seems like it would panic on shutdown, I think breaking out of the loop is better. That said, I wasn't able to observe a panic.